File: api/src/services/post.ts
import { PostMediaType, PostVisibility, ReactionType } from "@prisma/client";
import { prisma } from "../prisma/client";

export type PostMediaInput = {
    type?: PostMediaType;
    url: string;
    thumbnail?: string | null;
    metadata?: any;
};

export class PostService {
    public static async listPosts(filter?: { authorId?: string; tag?: string }) {
        const where: any = {};
        if (filter?.authorId) {
            where.authorId = filter.authorId;
        }
        if (filter?.tag) {
            where.tags = { has: filter.tag };
        }

        const posts = await prisma.post.findMany({
            where,
            orderBy: { createdAt: "desc" },
            include: {
                media: true,
                author: {
                    select: { id: true, name: true, profile_picture: true },
                },
                _count: { select: { reactions: true, comments: true } },
            },
        });

        return { posts };
    }

    public static async getPostById(id: string) {
        const post = await prisma.post.findUnique({
            where: { id },
            include: {
                media: true,
                reactions: true,
                author: {
                    select: { id: true, name: true, profile_picture: true },
                },
                comments: {
                    where: { parentCommentId: null },
                    orderBy: { createdAt: "desc" },
                    include: {
                        author: {
                            select: { id: true, name: true, profile_picture: true },
                        },
                        replies: {
                            orderBy: { createdAt: "asc" },
                            include: {
                                author: {
                                    select: { id: true, name: true, profile_picture: true },
                                },
                            },
                        },
                    },
                },
                _count: { select: { reactions: true, comments: true } },
            },
        });

        return { post };
    }

    public static async getPostByUserId(id: string) {
        const posts = await prisma.post.findMany({
            where: { authorId: id },
            include: {
                media: true,
                reactions: true,
                author: {
                    select: { id: true, name: true, profile_picture: true },
                },
                comments: {
                    where: { parentCommentId: null },
                    orderBy: { createdAt: "desc" },
                    include: {
                        author: {
                            select: { id: true, name: true, profile_picture: true },
                        },
                        replies: {
                            orderBy: { createdAt: "asc" },
                            include: {
                                author: {
                                    select: { id: true, name: true, profile_picture: true },
                                },
                            },
                        },
                    },
                },
                _count: { select: { reactions: true, comments: true } },
            },
        });

        return { posts };
    }

    public static async createPost(params: {
        authorId: string;
        content?: string;
        visibility?: PostVisibility;
        tags?: string[];
        media?: PostMediaInput[];
    }) {
        const { authorId, content, visibility, tags = [], media = [] } = params;

        return await prisma.$transaction(async (tx) => {
            const post = await tx.post.create({
                data: {
                    authorId,
                    content,
                    visibility: visibility || PostVisibility.PUBLIC,
                    tags,
                },
            });

            const normalizedMedia = media
                .filter((item) => item && item.url)
                .map((item) => ({
                    postId: post.id,
                    url: item.url,
                    thumbnail: item.thumbnail ?? null,
                    metadata: item.metadata,
                    type: item.type || PostMediaType.IMAGE,
                }));

            if (normalizedMedia.length > 0) {
                await tx.postMedia.createMany({ data: normalizedMedia });
            }

            const postWithRelations = await tx.post.findUnique({
                where: { id: post.id },
                include: { media: true, author: { select: { id: true, name: true, profile_picture: true } }, _count: { select: { reactions: true, comments: true } } },
            });

            return { post: postWithRelations };
        });
    }

    public static async updatePost(id: string, params: {
        content?: string;
        visibility?: PostVisibility;
        tags?: string[];
        media?: PostMediaInput[];
    }) {
        const { content, visibility, tags, media } = params;

        return await prisma.$transaction(async (tx) => {
            const existing = await tx.post.findUnique({ where: { id } });
            if (!existing) throw new Error("Post not found");

            await tx.post.update({
                where: { id },
                data: {
                    content: content ?? existing.content,
                    visibility: visibility ?? existing.visibility,
                    tags: tags ?? existing.tags,
                },
            });

            if (media) {
                await tx.postMedia.deleteMany({ where: { postId: id } });

                const normalizedMedia = media
                    .filter((item) => item && item.url)
                    .map((item) => ({
                        postId: id,
                        url: item.url,
                        thumbnail: item.thumbnail ?? null,
                        metadata: item.metadata,
                        type: item.type || PostMediaType.IMAGE,
                    }));

                if (normalizedMedia.length > 0) {
                    await tx.postMedia.createMany({ data: normalizedMedia });
                }
            }

            const postWithRelations = await tx.post.findUnique({
                where: { id },
                include: { media: true, author: { select: { id: true, name: true, profile_picture: true } }, _count: { select: { reactions: true, comments: true } } },
            });

            return { post: postWithRelations };
        });
    }

    public static async toggleReaction(postId: string, userId: string, type: ReactionType) {
        return await prisma.$transaction(async (tx) => {
            const existingPost = await tx.post.findUnique({ where: { id: postId } });
            if (!existingPost) {
                throw new Error("Post not found");
            }

            const existingReaction = await tx.postReaction.findUnique({
                where: { postId_userId: { postId, userId } },
            });

            if (existingReaction) {
                if (existingReaction.type === type) {
                    await tx.postReaction.delete({ where: { postId_userId: { postId, userId } } });
                    await tx.post.update({
                        where: { id: postId },
                        data: { reactionCount: { decrement: 1 } },
                    });
                    return { action: "removed" as const, reaction: null };
                }

                const reaction = await tx.postReaction.update({
                    where: { postId_userId: { postId, userId } },
                    data: { type },
                });

                return { action: "updated" as const, reaction };
            }

            const reaction = await tx.postReaction.create({ data: { postId, userId, type } });
            await tx.post.update({
                where: { id: postId },
                data: { reactionCount: { increment: 1 } },
            });

            return { action: "added" as const, reaction };
        });
    }

    public static async deleteReaction(postId: string, userId: string) {
        return await prisma.$transaction(async (tx) => {
            const existing = await tx.postReaction.findUnique({ where: { postId_userId: { postId, userId } } });
            if (!existing) return { deleted: false };

            await tx.postReaction.delete({ where: { postId_userId: { postId, userId } } });
            await tx.post.update({ where: { id: postId }, data: { reactionCount: { decrement: 1 } } });
            return { deleted: true };
        });
    }

    public static async addComment(postId: string, userId: string, content: string, parentCommentId?: string) {
        return await prisma.$transaction(async (tx) => {
            const post = await tx.post.findUnique({ where: { id: postId } });
            if (!post) {
                throw new Error("Post not found");
            }

            const comment = await tx.postComment.create({
                data: {
                    postId,
                    userId,
                    content,
                    parentCommentId: parentCommentId || null,
                },
                include: {
                    author: { select: { id: true, name: true, profile_picture: true } },
                    replies: true,
                },
            });

            await tx.post.update({
                where: { id: postId },
                data: { commentCount: { increment: 1 } },
            });

            return { comment };
        });
    }

    public static async updateComment(commentId: string, userId: string, content: string) {
        return await prisma.$transaction(async (tx) => {
            const existing = await tx.postComment.findUnique({ where: { id: commentId } });
            if (!existing) throw new Error("Comment not found");
            if (existing.userId !== userId) throw new Error("Forbidden");

            const comment = await tx.postComment.update({ where: { id: commentId }, data: { content, isEdited: true } });
            return { comment };
        });
    }

    public static async deleteComment(commentId: string, allowAnyUser = false, userId?: string) {
        return await prisma.$transaction(async (tx) => {
            const comment = await tx.postComment.findUnique({ where: { id: commentId } });
            if (!comment) throw new Error("Comment not found");
            if (!allowAnyUser && userId && comment.userId !== userId) throw new Error("Forbidden");

            const toDelete = await tx.postComment.findMany({
                where: {
                    OR: [{ id: commentId }, { parentCommentId: commentId }],
                },
                select: { id: true },
            });

            const deletedCount = toDelete.length;

            await tx.postComment.deleteMany({
                where: {
                    OR: [{ id: commentId }, { parentCommentId: commentId }],
                },
            });

            await tx.post.update({
                where: { id: comment.postId },
                data: { commentCount: { decrement: deletedCount } },
            });

            return { deletedCount };
        });
    }

    public static async deletePost(postId: string) {
        await prisma.$transaction(async (tx) => {
            await tx.postReaction.deleteMany({ where: { postId } });
            await tx.postComment.deleteMany({ where: { postId } });
            await tx.postMedia.deleteMany({ where: { postId } });
            await tx.post.delete({ where: { id: postId } });
        });
    }
}

File: mobile/lib/models/post.dart
import 'user.dart';

enum PostMediaType { IMAGE, VIDEO }

class PostMedia {
  final String id;
  final String url;
  final String type;
  final String? thumbnail;

  PostMedia({
    required this.id,
    required this.url,
    required this.type,
    this.thumbnail,
  });

  factory PostMedia.fromJson(Map<String, dynamic> json) {
    return PostMedia(
      id: json['id'],
      url: json['url'],
      type: json['type'],
      thumbnail: json['thumbnail'],
    );
  }
}

class PostCount {
  final int reactions;
  final int comments;

  PostCount({required this.reactions, required this.comments});

  factory PostCount.fromJson(Map<String, dynamic> json) {
    return PostCount(
      reactions: json['reactions'] ?? 0,
      comments: json['comments'] ?? 0,
    );
  }
}

class PostReaction {
  final String userId;
  final String type;

  PostReaction({required this.userId, required this.type});

  factory PostReaction.fromJson(Map<String, dynamic> json) {
    final userId =
        json['userId'] ??
        json['user_id'] ??
        json['user']?['id'] ??
        json['user']?['userId'];
    return PostReaction(
      userId: userId?.toString() ?? '',
      type: json['type']?.toString() ?? '',
    );
  }
}

class Post {
  final String id;
  final String? content;
  final String authorId;
  final DateTime createdAt;
  final List<PostMedia> media;
  final List<String> tags;
  final User author;
  final PostCount counts;
  final List<String> reactionUserIds;
  final List<PostReaction> reactions;

  Post({
    required this.id,
    this.content,
    required this.authorId,
    required this.createdAt,
    required this.media,
    required this.tags,
    required this.author,
    required this.counts,
    List<String>? reactionUserIds,
    List<PostReaction>? reactions,
  }) : reactionUserIds = reactionUserIds ?? const [],
       reactions = reactions ?? const [];

  factory Post.fromJson(Map<String, dynamic> json) {
    final parsedReactions = _parseReactions(json);
    return Post(
      id: json['id'],
      content: json['content'],
      authorId: json['authorId'],
      createdAt: DateTime.parse(json['createdAt']),
      media:
          (json['media'] as List<dynamic>?)
              ?.map((e) => PostMedia.fromJson(e))
              .toList() ??
          [],
      tags:
          (json['tags'] as List<dynamic>?)?.map((e) => e.toString()).toList() ??
          [],
      author: User.fromJson(json['author']),
      counts: PostCount.fromJson(json['_count'] ?? {}),
      reactionUserIds: _parseReactionUserIds(json, parsedReactions),
      reactions: parsedReactions,
    );
  }

  static List<PostReaction> _parseReactions(Map<String, dynamic> json) {
    final reactions = json['reactions'];
    if (reactions is! List) return const [];
    return reactions
        .whereType<Map>()
        .map((item) => PostReaction.fromJson(Map<String, dynamic>.from(item)))
        .where((r) => r.userId.isNotEmpty && r.type.isNotEmpty)
        .toList();
  }

  static List<String> _parseReactionUserIds(
    Map<String, dynamic> json,
    List<PostReaction> parsedReactions,
  ) {
    if (parsedReactions.isNotEmpty) {
      return parsedReactions.map((r) => r.userId).toList();
    }

    final reactions = json['reactions'];
    if (reactions is! List) return [];

    return reactions
        .map((item) {
          if (item is Map<String, dynamic>) {
            final userId =
                item['userId'] ??
                item['user_id'] ??
                item['user']?['id'] ??
                item['user']?['userId'];
            if (userId != null) return userId.toString();
          } else if (item != null) {
            return item.toString();
          }
          return null;
        })
        .whereType<String>()
        .where((id) => id.isNotEmpty)
        .toList();
  }
}
